File: Data Structures & Algorithms/sort-an-array/submission-1.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        return self.merge_sort(nums)

    def merge_sort(self, nums: List[int]) -> List[int]:
        if len(nums) == 1:
            return nums
        
        left = nums[:len(nums) // 2]
        arr1 = self.merge_sort(left)
        right = nums[len(nums) // 2:]
        arr2 = self.merge_sort(right)

        return self.merge(arr1, arr2)
    
    def merge(self, arr1: List[int], arr2: List[int]) -> List[int]:
        res = []
        i = j = 0

        while (i < len(arr1)) and (j < len(arr2)):
            if arr1[i] < arr2[j]:
                res.append(arr1[i])
                i += 1
            else:
                res.append(arr2[j])
                j += 1
        
        while i < len(arr1):
            res.append(arr1[i])
            i += 1
        
        while j < len(arr2):
            res.append(arr2[j])
            j += 1
        
        return res
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        int[] left = new int[length];
        int[] right = new int[length];
        
        left[0] = nums[0];
        int endI = length - 1;
        right[endI] = nums[endI];
        for (int i = 1; i < length; i++) {
            endI--;
            left[i] = nums[i] * left[i - 1];
            right[endI] = nums[endI] * right[endI + 1];
        }
        
        int[] output = new int[length];
        output[0] = right[1];
        output[length - 1] = left[length - 2];
        for (int i = 1; i < length - 1; i++) {
            output[i] = left[i - 1] * right[i + 1];
        }
        return output;
    }
}  
